src/main/java/net/minecraftforge/fml/common/asm/transformers/SoundEngineFixTransformer.java

```diff
@@ -0,0 +1,137 @@
+            {
+                if (m.name.equals("removeSource") && m.desc.equals("(Ljava/lang/String;)V"))
+                {
+                    method = m;
+                    break;
+                }
+            }
+            if (method == null)
+                throw new RuntimeException("Error processing " + transformedName + " - no removeSource method found");
+
+            for (Iterator<AbstractInsnNode> iterator = method.instructions.iterator(); iterator.hasNext();)
+            {
+                AbstractInsnNode insn = iterator.next();
+                if (insn instanceof MethodInsnNode && ((MethodInsnNode) insn).owner.equals("paulscode/sound/Source")
+                        && ((MethodInsnNode) insn).name.equals("cleanup"))
+                {
+                    LabelNode after = (LabelNode) insn.getNext();
+
+                    AbstractInsnNode beginning = insn.getPrevious();
+
+                    int varIndex = ((VarInsnNode) beginning).var;
+
+                    method.instructions.insertBefore(beginning, new VarInsnNode(Opcodes.ALOAD, varIndex));
+                    method.instructions.insertBefore(beginning, new FieldInsnNode(Opcodes.GETFIELD, "paulscode/sound/Source", "toStream", "Z"));
+                    LabelNode elseNode = new LabelNode();
+                    method.instructions.insertBefore(beginning, new JumpInsnNode(Opcodes.IFEQ, elseNode));
+
+                    method.instructions.insertBefore(beginning, new VarInsnNode(Opcodes.ALOAD, varIndex));
+                    method.instructions.insertBefore(beginning, new InsnNode(Opcodes.ICONST_1));
+                    method.instructions.insertBefore(beginning, new FieldInsnNode(Opcodes.PUTFIELD, "paulscode/sound/Source", "removed", "Z"));
+
+                    method.instructions.insertBefore(beginning, new JumpInsnNode(Opcodes.GOTO, after));
+
+                    method.instructions.insertBefore(beginning, elseNode);
+                    break;
+                }
+            }
+
+            ClassWriter writer = new ClassWriter(ClassWriter.COMPUTE_MAXS);
+            classNode.accept(writer);
+            return writer.toByteArray();
+        }
+        else if (transformedName.equals("paulscode.sound.StreamThread"))
+        {
+            ClassNode classNode = new ClassNode();
+            ClassReader classReader = new ClassReader(basicClass);
+            classReader.accept(classNode, 0);
+
+            MethodNode method = null;
+            for (MethodNode m : classNode.methods)
+            {
+                if (m.name.equals("run") && m.desc.equals("()V"))
+                {
+                    method = m;
+                    break;
+                }
+            }
+            if (method == null)
+                throw new RuntimeException("Error processing " + transformedName + " - no run method found");
+
+            for (Iterator<AbstractInsnNode> iterator = method.instructions.iterator(); iterator.hasNext();)
```

Comment:
Same issue with concurrent modification in this loop. Also why not use a foreach loop?

Reply:
The field `instructions` (type `InsnList`) is not Iterable.